src/packet/ipv4.rs

```diff
@@ -39,29 +41,67 @@ pub struct Ipv4 {
     payload: Vec<u8>,
 }
 
-/// Calculates the checksum of an IPv4 packet
-pub fn checksum(packet: &Ipv4Packet) -> u16be {
-    use packet::Packet;
+/// Represents an IPv4 Packet
+///
+/// 0          8          16          24         32
+/// +---------------------------------------------+
+/// |                  source_ip                  |
+/// +----------------------+----------------------+
+/// |                destination_ip               |
+/// +-----------+----------+----------------------+
+/// |   zeros   | protocol | tcp/udp length       |
+/// +-----------+---------------------------------+
+#[packet]
+pub struct Ipv4PseudoHeader {
+    #[construct_with(u8, u8, u8, u8)]
+    source: Ipv4Addr,
+    #[construct_with(u8, u8, u8, u8)]
+    destination: Ipv4Addr,
+    zeros: u8,
+    #[construct_with(u8)]
+    next_level_protocol: IpNextHeaderProtocol,
+    inner_packet_length: u16be,
+    // just to avoid the compiler to complain
+    #[payload]
+    payload: Vec<u8>,
+}
 
-    let len = packet.get_header_length() as usize * 4;
-    let mut sum = 0u32;
-    let mut i = 0;
-    while i < len {
-        let word = (packet.packet()[i] as u32) << 8 | packet.packet()[i + 1] as u32;
-        sum = sum + word;
-        i = i + 2;
-    }
-    while sum >> 16 != 0 {
-        sum = (sum >> 16) + (sum & 0xFFFF);
+impl<'p> PseudoHeader for Ipv4Packet<'p> {
+
+    /// Return a PseudoHeader packet out of the packet header.
+    ///
+    /// The `inner_packet_length` is optional, since in case of UDP, it is taken from the length
+    /// field of the UDP header, whereas for TCP, it is computed from the header length.
+    fn get_pseudo_header(&self, inner_packet_length: Option<u32>) -> Vec<u8> {
```

Comment:
that function signature won't work for our PseudoHeader trait because it must return a value which is suitable for ipv4 and ipv6... what should we do instead? perhaps an enum type which is either Ipv4PseudoHeader or Ipv6PseudoHeader?